boa_engine/src/vm/graph.rs

```diff
@@ -0,0 +1,873 @@
+                Color::Yellow => "yellow".into(),
+                Color::Purple => "purple".into(),
+                Color::Color(color) => format!("#{color:X}"),
+            };
+            let style = match (edge.style, edge.type_) {
+                (EdgeStyle::Line, EdgeType::None) => "---",
+                (EdgeStyle::Line, EdgeType::Arrow) => "-->",
+                (EdgeStyle::Dotted | EdgeStyle::Dashed, EdgeType::None) => "-.-",
+                (EdgeStyle::Dotted | EdgeStyle::Dashed, EdgeType::Arrow) => "-.->",
+            };
+            result.push_str(&format!(
+                "  {}_i_{} {style}| {}| {}_i_{}\n",
+                self.label,
+                edge.from,
+                edge.label.as_deref().unwrap_or(""),
+                self.label,
+                edge.to,
+            ));
+
+            if !color.is_empty() {
+                result.push_str(&format!(
+                    "  linkStyle {} stroke:{}, stroke-width: 4px\n",
+                    index + 1,
+                    color
+                ));
+            }
+        }
+        for subgraph in &self.subgraphs {
+            subgraph.mermaid_format(result);
+        }
+        result.push_str("  end\n");
+    }
+}
+
+#[derive(Debug)]
+pub struct Graph {
+    subgraphs: Vec<SubGraph>,
+    direction: Direction,
+}
+
+impl Graph {
+    pub fn new(direction: Direction) -> Self {
+        Graph {
+            subgraphs: Vec::default(),
+            direction,
+        }
+    }
+
+    pub fn subgraph(&mut self, label: String) -> &mut SubGraph {
+        self.subgraphs.push(SubGraph::new(label));
+        let result = self
+            .subgraphs
+            .last_mut()
+            .expect("We just pushed a subgraph");
+        result.set_direction(self.direction);
+        result
+    }
+
+    pub fn to_graphviz_format(&self) -> String {
+        let mut result = String::new();
+        result += "digraph {\n";
+        result += "\tnode [shape=record];\n";
+
+        let rankdir = match self.direction {
+            Direction::TopToBottom => "TB",
+            Direction::BottomToTop => "BT",
+            Direction::LeftToRight => "LR",
+            Direction::RightToLeft => "RL",
+        };
+        result += &format!("\trankdir={rankdir};\n");
+
+        for subgraph in &self.subgraphs {
+            subgraph.graphviz_format(&mut result);
+        }
+        result += "}\n";
+        result
+    }
+
+    pub fn to_mermaid_format(&self) -> String {
+        let mut result = String::new();
+        let rankdir = match self.direction {
+            Direction::TopToBottom => "TD",
+            Direction::BottomToTop => "DT",
+            Direction::LeftToRight => "LR",
+            Direction::RightToLeft => "RL",
+        };
+        result += &format!("graph {}\n", rankdir);
+
+        for subgraph in &self.subgraphs {
+            subgraph.mermaid_format(&mut result);
+        }
+        result += "\n";
+        result
+    }
+}
+
+impl CodeBlock {
```

Comment:
We might want to separate this `impl` into a different file (or separate the rest into a different file), to keep file size reasonable.